### PR TITLE
Enable alarm clock functionality access from the 'wait' command

### PR DESCRIPTION
**Pull request by esotericist**
#### Summary
SUMMARY: Features "Enable alarm clock functionality access from the 'wait' command"

#### Purpose of change
Now that we have interruptible and resumable crafts, it'd be nice to have the ability to interrupt crafts on a player-set schedule.

#### Describe the solution
Extended the 'wait' command to, if the player has access to an alarm clock, ask the player if they want to set an alarm or if they want to wait as normal.

If they choose to set the alarm, they're given a variation on the wait UI to select a timepoint for when they want the alarm to go off. This sets the `effect_alarm_clock` effect on the player with an appropriate duration.

When that duration expires, if they still have access to an alarm clock, an interruption is fired. If they do not have access to an alarm clock, the event ends harmlessly (no sound is generated).

#### Describe alternatives you've considered
I originally wanted to have the alarm clock be activated in some fashion as per tools, but all of the available methods for doing that would have complicated a number of things. 

#### Additional context
The game has no notion of *which* alarm clock you used to set the alarm, which means if you set an alarm, drop the clock, move 20 tiles away to a second alarm, the event will fire from your new location. This is presently not avoidable in a reasonable fashion, but players are unlikely to be doing this.

image

image

image

image

image

image

### CODE REVIEW

File: src/handle_action.cpp [hunk 1]
```diff
@@ -655,12 +655,39 @@ static void smash()
...
     std::map<int, int> durations;
     uilist as_m;
+    player &u = g->u;
+    bool setting_alarm = false;
+
+    if( u.has_alarm_clock() ) {
+        setting_alarm = try_set_alarm();
+    }
+
+    const bool has_watch = ( u.has_watch() || setting_alarm );
```
Comment (Night-Pryanik) [inline src/handle_action.cpp]: Exterior brackets are redundant.

File: src/handle_action.cpp [hunk 2]
```diff
@@ -675,14 +702,24 @@ static void wait()
         durations[retval] = duration;
     };
 
-    add_menu_item( 1, '1', !has_watch ? _( "Wait 300 heartbeats" ) : "", MINUTES( 5 ) );
-    add_menu_item( 2, '2', !has_watch ? _( "Wait 1800 heartbeats" ) : "", MINUTES( 30 ) );
+    if( setting_alarm ) {
+
+        add_menu_item( 0, '0', _( "" ), MINUTES( 30 ) );
```
Comment (Night-Pryanik) [inline src/handle_action.cpp]: Is there a need for translating `""`?

File: src/handle_action.cpp (hunk 1 above)
Comment (esotericist) [reply src/handle_action.cpp]: Bad habits from java. Will fix.

File: src/handle_action.cpp (hunk 2 above)
Comment (esotericist) [reply src/handle_action.cpp]: Oversight from edits. I briefly had text there during testing and I was being neurotic about strings. Will fix.

**Comment by esotericist** [comment]
That's both items @Night-Pryanik  pointed out addressed.

**Comment by Fris0uman** [comment]
>which means if you set an alarm, drop the clock, move 20 tiles away to a second alarm, the event will fire from your new location. This is presently not avoidable in a reasonable fashion, but players are unlikely to be doing this.

I'm not usre it''s that unlikely, for exemple you could want to use the alarm as a distraction to go off later and draw zombies away from you.

**Comment by Zireael07** [comment]
@Fris0uman: I think that's a good point, but for another PR? This is a very welcome addition that shouldn't be held up by more future ideas.

**Comment by esotericist** [comment]
> > which means if you set an alarm, drop the clock, move 20 tiles away to a second alarm, the event will fire from your new location. This is presently not avoidable in a reasonable fashion, but players are unlikely to be doing this.
> 
> I'm not sure it''s that unlikely, for exemple you could want to use the alarm as a distraction to go off later and draw zombies away from you.

Yeah, I can see that, but on the timescale the alarm clock is meant to work on, as well as the comparatively low volume it has, there's other things that would be more useful for that.

That said, I might try to do something like this with the discrete alarm clock or an egg timer (both as non-worn items) in a future PR.

**Comment by ZhilkinSerg** [comment]
Jenkins rebuild

**Comment by esotericist** [comment]
Found an oversight in what happens if a player has an alarm clock set when attempting to sleep and wakes up at an unusual time (before or after the alarm), or if the player never slept.

**Pull request merged**